psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1791,597 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+    std::vector<std::vector<double>> Sai_next(num_atoms, std::vector<double>(4, 0.0));
+
+    // Population and width guesses, from get_mbis_params function
+    for (int atom = 0; atom < num_atoms; atom++) {
+        int atomic_num = (int) mol->Z(atom);        
+
+        for (int m = 0; m < mA[atom]; m++) {
+            std::tie(Nai[atom][m], Sai[atom][m]) = get_mbis_params(atomic_num, m);
+            Sai[atom][m] = 1.0/Sai[atom][m];
+            
+            if (print_output && debug >= 1) {
+                outfile->Printf("  INITIAL ATOM %d, SHELL %d, POP %8.5f, WIDTH %8.5f\n", atom+1, m+1, Nai[atom][m], Sai[atom][m]);
+            }
+            
+            Nai_next[atom][m] = Nai[atom][m];
+            Sai_next[atom][m] = Sai[atom][m];
+        }
+    }
+
+    // Promolecular and proatomic densities
+    std::vector<double> rho_0_points(total_points, 0.0);
+    std::vector<std::vector<double>> rho_a_0_points(num_atoms, std::vector<double>(total_points, 0.0));
+
+    // => Main Stockholder Loop <= //
+
+    int iter = 1;
+    bool is_converged = false;
+    double delta_rho_max_0;
+
+    if (print_output && debug >= 1) outfile->Printf("                     Delta D\n");
+    while (iter < max_iter) {
+
+        // Calculate proatom and promolecule density at all points
+#pragma omp parallel for
+        for (int point = 0; point < total_points; point++) {
+            rho_0_points[point] = 0.0;
+            for (int atom = 0; atom < num_atoms; atom++) {
+                rho_a_0_points[atom][point] = 0.0;
+                for (int m = 0; m < mA[atom]; m++) {
+                    rho_a_0_points[atom][point] += rho_ai_0(Nai[atom][m], Sai[atom][m], distances[atom][point]);
+                }
+                rho_0_points[point] += rho_a_0_points[atom][point];
+            }
+        }
+
+        // Self-consistent update of population and density
+#pragma omp parallel for
+        for (int atom = 0; atom < num_atoms; atom++) {
+            for (int m = 0; m < mA[atom]; m++) {
+
+                double sum_n = 0.0;
+                double sum_s = 0.0;
+
+                for (int point = 0; point < total_points; point++) {
+                    double rho_ai_0_point = rho_ai_0(Nai[atom][m], Sai[atom][m], distances[atom][point]);
+                    sum_n += weights[point] * rho[point] * rho_ai_0_point / rho_0_points[point];
+                    sum_s += weights[point] * distances[atom][point] * rho[point] * rho_ai_0_point / rho_0_points[point];
+                }
+
+                Nai_next[atom][m] = sum_n;
+                Sai_next[atom][m] = sum_s / (3 * Nai_next[atom][m]);
+            }
+        }
+
+        // Convergence check (Equation 20 in Verstraelen et al.)
+        std::vector<double> delta_rho_atoms_0(num_atoms, 0.0);
+#pragma omp parallel for
+        for (int atom = 0; atom < num_atoms; atom++) {
+
+            for (int point = 0; point < total_points; point++) {
```

Comment:
But that ruins the memory access.

Reply:
This is a tricky situation. Holger is definitely right that parallel scaling would be better if the loops were switched. Unfortunately, the `atom` loop has to be outside of the `point` loop here because we're summing over all points to compute an atomic error. In other places the opposite is true, and the `point` loop has to be outer.

As Susi points out, storing data as `[atom][point]` is bad for memory access if the outer for loop is over `point`. Since we're forced to use a mix of loop orderings (sometime `atom` is outer, other times `point`), it's not clear which memory pattern is better, so I timed both patterns on a large system:

|                        | 1 core time (s) | 6 core time (s)  |
| ------------- |:---------:|:-----:|
| [atom][point] | 274 | 62.7 |
| [point][atom] | 300 |  65.3 |